### ISSUE TITLE
Computational efficiency

### ISSUE THREAD

**Issue by ShilinC** [microsoft/O-CNN#101]
Hi,

I'm wondering if all these different versions of the code (caffe, TF, Pytorch) can produce similar speed gain compared to the standard 3D CNN (voxel CNN)? In the slides I can see O-CNN is much more efficient in terms of the measured time (O(N^2) compared to O(N^3)). Did you verify this on all these 3 learning platforms? Or does the current code allow me to run this comparison directly?

I can understand that O-CNN is faster given that the number of convolution operations is smaller than a standard 3D CNN and it is much more memory efficient. So I guess the only overhead is the neighboring octant search (so I guess you optimize this part by implementing shuffle keys on the GPU very efficiently?). This is because previously I tried some other quadtree based neural nets and sometimes they even consumes larger time than a standard 2D CNN on some tasks (I guess it may because of the suboptimal GPU implementation and standard CNN is optimized to leverage the parallelization), so I'm curious on if you optimized anything implementation-wise in order to make it efficient (if so, could you point out which part of the code corresponds to the speed optimization?). Please correct me if I'm wrong. Thanks!!!

**Comment by wang-ps** [comment]
The three versions of code (caffe, TF, Pytorch) reuse the same underlying C++/CUDA functions, so the running speed is the same.  Recently, I tried the [O-CNN base LeNet with TF](https://github.com/microsoft/O-CNN/blob/master/docs/classification.md#o-cnn-on-tensorflow), and the average time of one forward+backward iteration is roughly 10ms on Geforce 2080Ti with batch 32.  And I find that with TF the memory consumption is also greatly reduced especially when the network is very deep ( due to TF's memory management mechanism). For example, the [O-CNN base Resnet] (https://github.com/microsoft/O-CNN/blob/master/docs/classification.md#o-cnn-on-tensorflow) only takes less than 3GB memroy with TF, whereas it takes about 9GB+ memory with caffe.   

Originally, we used the caffe-based implementation of 3D Voxel CNN to do the speed/memory comparison. You can checkout the official caffe code (with commits back to the winter of 2016) to reproduce the experiments. And I did not do the comparison under TF or Pytorch. The released code does not contain the 3D Voxel CNN.

When doing convolution, yes, the overhead is neighbor searching. In the first version of the implementation, I used Hash Table to speed up (See Section 3.2 on Page 5 of our paper). Considering the kernel size of Conv of often small than 3*3*3 in modern convolution architectures (such as VGG or ResNet), we can pre-compute the 3*3*3 neighbor indices to further speed up the code (See the `calc_neigh_cpu` and `calc_neigh_gpu` [here](https://github.com/microsoft/O-CNN/blob/4fdcfe249e61b4ead162e05066e6ee34c479bf02/octree/octree/octree_nn.h#L122) and `octree2col_cpu` and  `octree2col_gpu` function [here](https://github.com/microsoft/O-CNN/blob/4fdcfe249e61b4ead162e05066e6ee34c479bf02/octree/octree/octree_nn.h#L81)) 

For other CNN operations such as BatchNorm or nonlinear activations, the speed is directly related to the number of voxels involved. For these operations, the O-CNN has no overhead. For pooling, since the octree nodes are sorted with the shuffled key, the octree nodes belong to one parent one is contiguous, the speed is even faster than the Voxel-base memory layout.

**Comment by wang-ps** [comment]
Yes, you are right.

SparseConvNet is less efficient when the network becomes very deep. It's because the active sites keep growing. 

In principle, the Submanifold Convolution is the same as our O-CNN, i.e. the CNN operations only occur in the active voxels. The only difference is that we use an octree data structure to organize the sparse active/non-empty voexls, whereas Submanifold Sparse Convolution used a hash table as the data structure.   

We did not do the comparison with Subminifold SparseConvNet in our original paper,  since when submitting our paper to Siggraph 2017 (at the end of the year 2016), the authors of Submanifold CNN have not posted their paper to the Arxiv. And its improved version was published in CVPR 2018. Unfortunately, they missed our paper in their citation.

According to my preliminary comparison a few months ago, our implementation runs faster.

**Comment by ShilinC** [comment]
Thanks a lot! Great information.
